language/bytecode_verifier/src/type_memory_safety.rs

```diff
@@ -51,8 +51,24 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
         let locals_signature_view = function_definition_view.locals_signature();
         let function_signature_view = function_definition_view.signature();
         let mut locals = BTreeMap::new();
+        if function_signature_view.arg_count() > locals_signature_view.len() {
+            return vec![VMStatus::new(StatusCode::RANGE_OUT_OF_BOUNDS)
+                .with_message("Fewer locals than parameters".to_string())];
+        }
+        let mut errors = vec![];
         for (arg_idx, arg_type_view) in function_signature_view.arg_tokens().enumerate() {
-            if arg_type_view.is_reference() {
+            if arg_type_view.as_inner()
+                != locals_signature_view
+                    .token_at(arg_idx as LocalIndex)
+                    .as_inner()
+            {
+                errors.push(
+                    VMStatus::new(StatusCode::TYPE_MISMATCH).with_message(format!(
+                        "Type mismatch at index {} between parameter and local",
+                        arg_idx
+                    )),
+                );
+            } else if arg_type_view.is_reference() {
```

Comment:
This `else` feels a bit brittle to me. I know that you don't "need" this code to be run if there are errors. But if something changes in the future this could lead to bugs.

Reply:
I have split the checking out into its own code.  So this loop remains just as before and will be reached only if the checking succeeds.